File: CPlus/designLinkedList.cpp
class MyLinkedList {

private:
    struct node{
        int val;
        node *next;
    }; 
    
    node *head;
    node *tail;
    int size;
    
    
public:
    /** Initialize your data structure here. */
    MyLinkedList() {
        head=nullptr;
        tail=nullptr;
        size=0;
    }
    
    /** Get the value of the index-th node in the linked list. If the index is invalid, return -1. */
    int get(int index) {
        if(index>=size || index<0) return -1;
        node* cur=head;
        for (int i=0;i<index;i++){
            cur=cur->next;
        }
        
        return cur->val;
        
    }
    
    /** Add a node of value val before the first element of the linked list. After the insertion, the new node will be the first node of the linked list. */
    void addAtHead(int val) {
        node *temp = new node();
        temp->val = val;
        temp->next = head;
        head = temp;
        if(size==0) tail = temp;
        ++size;
    }
    
    /** Append a node of value val to the last element of the linked list. */
    void addAtTail(int val) {
        node *temp=new Node();
        temp->val=val;
        if (size==0){
            tail = temp;
            head = temp;
        }
        tail->next=temp;
        tail=temp;
        size++;
    }
    
    /** Add a node of value val before the index-th node in the linked list. If index equals to the length of linked list, the node will be appended to the end of linked list. If index is greater than the length, the node will not be inserted. */
    void addAtIndex(int index, int val) {
        if (index > size)
            return;
        
        if (index == 0) {
            addAtHead(val);
            return;
        }
        if (index == size) {
            addAtTail(val);
            return;
        }
        node *p = head;
        for (int i = 0; i < index-1; i++) {
            p = p->next;
        }
        node *q = new node();
        q->val=val;
        q->next = p->next;
        p->next = q;
        size++;
    }
    
    /** Delete the index-th node in the linked list, if the index is valid. */
    void deleteAtIndex(int index) {
                if (index >= size)
            return;
        
        node *p = head;
        if (index == 0) {
            head = head->next;
            delete p;
            size--;
            return;
        }

        for (int i = 0; i < index-1; i++) {
            p = p->next;
        }
        node *q = p->next;
        p->next = q->next;
        if (tail == q) tail = p;
        delete q;
        size--;
    }
};

File: Python/designLinkedList.py
class Node:
    def __init__(self, val, _next=None):
        self.val = val
        self.next = _next

class MyLinkedList(object):
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.head=None
        self.tail=None
        self.size=0
        

    def get(self, index):
        """
        Get the value of the index-th node in the linked list. If the index is invalid, return -1.
        :type index: int
        :rtype: int
        """
        if index<0 or index>=self.size:
            return -1
        
        current=Node(0,self.head)
        for i in range(index+1):
            current=current.next
        return current.val
            

    def addAtHead(self, val):
        """
        Add a node of value val before the first element of the linked list. After the insertion, the new node will be the first node of the linked list.
        :type val: int
        :rtype: None
        """
        toAdd=Node(val,self.head)
        self.head=toAdd
        if self.size==0:
            self.tail=toAdd
        
        self.size+=1
        
        

    def addAtTail(self, val):
        """
        Append a node of value val to the last element of the linked list.
        :type val: int
        :rtype: None
        """
        toAdd=Node(val)
        if self.size==0:
            self.head=self.tail=toAdd
        else:
            self.tail.next=toAdd
            self.tail=toAdd     
        self.size+=1
        
    def addAtIndex(self, index, val):
        """
        Add a node of value val before the index-th node in the linked list. If index equals to the length of linked list, the node will be appended to the end of linked list. If index is greater than the length, the node will not be inserted.
        :type index: int
        :type val: int
        :rtype: None
        """
        if index<0 or index>self.size: return
        if index==0: return self.addAtHead(val)
        if index==self.size: return self.addAtTail(val)
        
        current=Node(0,self.head)
        for i in range(index):
            current=current.next
        toAdd=Node(val,current.next)
        current.next=toAdd
        self.size+=1
        
        
        
        

    def deleteAtIndex(self, index):
        """
        Delete the index-th node in the linked list, if the index is valid.
        :type index: int
        :rtype: None
        """
        if index < 0 or index >= self.size: return
        current=Node(0,self.head)
        for i in range(index):
            current=current.next
        current.next=current.next.next
        if index==0:
            self.head=current.next
        
        if index==self.size-1:
            self.tail=current
        
        self.size-=1